### DISCUSSION TITLE
Why does the SQLAlchemy code distinguish between bound=object and bound=Any?

### DISCUSSION THREAD

**Discussion by pamelafox** [sqlalchemy/sqlalchemy#9835] [Usage Questions] [answered]
I'm working on a Flask-SQLAlchemy PR to improve the typing, and am basing some types on how SQLAlchemy does it. I'm puzzled over this:

```
# I would have preferred this were bound=object however it seems
# to not travel in all situations when defined in that way.
_O = TypeVar("_O", bound=Any)
"""The 'ORM mapped object' type.
"""

_OO = TypeVar("_OO", bound=object)
"""The 'ORM mapped object, that's definitely object' type.
"""
```

From 
https://github.com/sqlalchemy/sqlalchemy/blob/2dced78f33010d4a6a6276f4e8c5665bd08d01c0/lib/sqlalchemy/orm/_typing.py

In what situations does _OO not work? It seems that _O is used in most of the signatures.

Thanks!

**Comment by zzzeek** [comment, marked as answer]
I recall coming across this difference but i dont recall what the error was, it was also an older version of mypy.     I will say that typing tools are very liberal when they see `Any` in a different way when they see any other type, even `object` which is the base type of literally everything in the Python interpreter.

Let's try right now to see if theres any difference; first apply this diff:

```diff
diff --git a/lib/sqlalchemy/orm/_typing.py b/lib/sqlalchemy/orm/_typing.py
index 36dc6ddb9c..c433d98a72 100644
--- a/lib/sqlalchemy/orm/_typing.py
+++ b/lib/sqlalchemy/orm/_typing.py
@@ -55,7 +55,7 @@ _T_co = TypeVar("_T_co", bound=Any, covariant=True)
 
 # I would have preferred this were bound=object however it seems
 # to not travel in all situations when defined in that way.
-_O = TypeVar("_O", bound=Any)
+_O = TypeVar("_O", bound=object)
 """The 'ORM mapped object' type.
 
 """

````

then let's run the pep-484 suite - here's the errors we get:

```
$ tox -e pep484
# ...
lib/sqlalchemy/orm/state.py:604: error: Incompatible types in assignment (expression has type "List[Any]", target has type "Union[Set[str], Dict[str, Any], Type[_O], _O, None]")  [assignment]
lib/sqlalchemy/orm/state.py:606: error: Incompatible types in assignment (expression has type "_SerializeManager", target has type "Union[Set[str], Dict[str, Any], Type[_O], _O, None]")  [assignment]
lib/sqlalchemy/orm/util.py:837: error: Item "object" of "Union[_O, AliasedClass[_O]]" has no attribute "__name__"  [union-attr]
lib/sqlalchemy/orm/util.py:1296: error: Item "object" of "Union[_O, AliasedClass[_O]]" has no attribute "__name__"  [union-attr]
lib/sqlalchemy/orm/util.py:1304: error: Item "object" of "Union[_O, AliasedClass[_O]]" has no attribute "__name__"  [union-attr]
lib/sqlalchemy/orm/decl_api.py:1607: error: "Type[_O]" has no attribute "_sa_apply_dc_transforms"  [attr-defined]
lib/sqlalchemy/ext/asyncio/session.py:593: error: Incompatible return value type (got "Optional[_O]", expected "Optional[_O]")  [return-value]

```

reverting that patch and running pep-484 again the errors go away.

**Reply by pamelafox** [reply]
Thanks @zzzeek! I'm new to typing in Python, so I also asked generally about bound= in the Python forum, if you're curious to see that discussion:
https://discuss.python.org/t/differences-in-bound-object-vs-bound-any/27052

For the three errors about "__name__", I wonder if a `TypeVar` with `bound=type` would be appropriate there? I think those are cases where the variable is a class that's an instance of `type`, and even the SQLAlchemy metaclasses like _DynamicAttributesType inherit from `type`. This passes mypy:

```
from typing import Any, TypeVar

A = TypeVar('A', bound=Any)
O = TypeVar('O', bound=type)


def thing1(a: A):
    a.__name__


def thing2(a: O):
    a.__name__
```

This is just musing based on what little I know of typing, and me trying to understand it better. It's also not a big deal, so I appreciate your response. :)

**Reply by zzzeek** [reply]
I think you're correct, the ones that refer to ``__name__`` are actually referring to "instance of types", the way `AliasedClass` is mostly like a `Type` object;  here that would be written as:

`Union[Type[_O], AliasedClass[_O]]`

that is, `_O` could be `object` as we want it to be.

anyway yeah we could likely fix this `_O` thing since that's very few errors

**Reply by zzzeek** [reply]
see https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4637 where I have all of these errors resolved.  thanks for the tip!

**Comment by CaselIT** [comment]
Last one is definitely the best error :/
